src/internal/connector/graph/middleware.go

```diff
@@ -122,42 +124,61 @@ func (mw *LoggingMiddleware) Intercept(
 		return resp, err
 	}
 
-	ctx = clues.Add(ctx, "status", resp.Status, "statusCode", resp.StatusCode)
-	log := logger.Ctx(ctx)
+	ctx = clues.Add(
+		ctx,
+		"status", resp.Status,
+		"statusCode", resp.StatusCode,
+		"content_len", resp.ContentLength)
 
-	// Return immediately if the response is good (2xx).
-	// If api logging is toggled, log a body-less dump of the request/resp.
-	if (resp.StatusCode / 100) == 2 {
-		if logger.DebugAPIFV || os.Getenv(log2xxGraphRequestsEnvKey) != "" {
-			log.Debugw("2xx graph api resp", "response", getRespDump(ctx, resp, os.Getenv(log2xxGraphResponseEnvKey) != ""))
-		}
+	var (
+		log       = logger.Ctx(ctx)
+		respClass = resp.StatusCode / 100
+		logExtra  = logger.DebugAPIFV || os.Getenv(logGraphRequestsEnvKey) != ""
+	)
+
+	// special case: always info log 429 responses
+	if resp.StatusCode == http.StatusTooManyRequests {
+		log.Infow(
+			"graph api throttling",
+			"limit", resp.Header.Get(rateLimitHeader),
+			"remaining", resp.Header.Get(rateRemainingHeader),
+			"reset", resp.Header.Get(rateResetHeader),
+			"retry-after", resp.Header.Get(retryAfterHeader))
 
 		return resp, err
 	}
 
-	// Log errors according to api debugging configurations.
-	// When debugging is toggled, every non-2xx is recorded with a response dump.
-	// Otherwise, throttling cases and other non-2xx responses are logged
-	// with a slimmer reference for telemetry/supportability purposes.
-	if logger.DebugAPIFV || os.Getenv(logGraphRequestsEnvKey) != "" {
-		log.Errorw("non-2xx graph api response", "response", getRespDump(ctx, resp, true))
+	// special case: always dump status-400-bad-request
+	if resp.StatusCode == http.StatusBadRequest {
+		log.With("response", getRespDump(ctx, resp, true)).
+			Error("graph api error: " + resp.Status)
+
 		return resp, err
 	}
 
-	msg := fmt.Sprintf("graph api error: %s", resp.Status)
+	// Log api calls according to api debugging configurations.
+	switch respClass {
+	case 2:
+		if logExtra {
+			// only dump the body if it's under a size limit.  We don't want to copy gigs into memory for a log.
+			dump := getRespDump(ctx, resp, os.Getenv(log2xxGraphResponseEnvKey) != "" && resp.ContentLength < logMBLimit)
```

Comment:
1Mb per 2xx response still seems like a lot? With 10k such responses, we would reach 10GB ( over a ~10 min period).

Reply:
I guess the average response would be much smaller.

Reply:
I'm honestly not sure there's any cases where dumping a body on a 2xx helps us at all.  This is one step towards memory management before removing it altogether.  I think we'd be smarter to remove it, but I'm going to keep it in for now just to be sure.

Reply:
Fair. I agree this is the best approach right now